File: c/dsa/reverse-linked-list/reverse_linked_list.c
// Reverse a singly linked list (iterative, three-pointer).
//
// Before: 1 -> 2 -> 3 -> 4 -> 5
// After:  5 -> 4 -> 3 -> 2 -> 1
//
// Build:  cc -O2 -Wall -Wextra -o reverse reverse_linked_list.c
// Run:    ./reverse

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int val;
    struct Node *next;
} Node;

static Node *new_node(int val, Node *next) {
    Node *n = malloc(sizeof(*n));
    n->val = val;
    n->next = next;
    return n;
}

static Node *reverse(Node *head) {
    Node *prev = NULL, *curr = head;
    while (curr) {
        Node *next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

static int to_array(Node *head, int *out, int cap) {
    int n = 0;
    while (head && n < cap) {
        out[n++] = head->val;
        head = head->next;
    }
    return n;
}

static Node *from_array(const int *xs, int n) {
    Node *head = NULL;
    for (int i = n - 1; i >= 0; i--) head = new_node(xs[i], head);
    return head;
}

static void free_list(Node *head) {
    while (head) {
        Node *next = head->next;
        free(head);
        head = next;
    }
}

static void print_ints(const char *label, const int *xs, int n) {
    printf("%s", label);
    for (int i = 0; i < n; i++) printf("%d ", xs[i]);
    printf("\n");
}

int main(void) {
    const int input[] = {1, 2, 3, 4, 5};
    const int expected[] = {5, 4, 3, 2, 1};
    const int n = (int)(sizeof(input) / sizeof(*input));

    Node *head = from_array(input, n);
    int buf[16];

    int len = to_array(head, buf, 16);
    print_ints("Before: ", buf, len);

    head = reverse(head);

    len = to_array(head, buf, 16);
    print_ints("After:  ", buf, len);

    assert(len == n);
    for (int i = 0; i < n; i++) assert(buf[i] == expected[i]);
    printf("OK — reversed list matches expected\n");

    free_list(head);
    return 0;
}

File: zig/dsa/reverse-linked-list/reverse_linked_list.zig
// Reverse a singly linked list (iterative, three-pointer).
//
// Before: 1 -> 2 -> 3 -> 4 -> 5
// After:  5 -> 4 -> 3 -> 2 -> 1
//
// Build/run:  zig run reverse_linked_list.zig

const std = @import("std");

const Node = struct {
    val: i32,
    next: ?*Node = null,
};

fn newNode(alloc: std.mem.Allocator, val: i32, next: ?*Node) !*Node {
    const n = try alloc.create(Node);
    n.* = .{ .val = val, .next = next };
    return n;
}

fn reverse(head: ?*Node) ?*Node {
    var prev: ?*Node = null;
    var curr = head;
    while (curr) |c| {
        const next = c.next;
        c.next = prev;
        prev = c;
        curr = next;
    }
    return prev;
}

fn toArray(head: ?*Node, out: []i32) usize {
    var count: usize = 0;
    var curr = head;
    while (curr) |c| {
        out[count] = c.val;
        count += 1;
        curr = c.next;
    }
    return count;
}

fn fromArray(alloc: std.mem.Allocator, xs: []const i32) !?*Node {
    var head: ?*Node = null;
    var i = xs.len;
    while (i > 0) {
        i -= 1;
        head = try newNode(alloc, xs[i], head);
    }
    return head;
}

fn printInts(label: []const u8, xs: []const i32) void {
    std.debug.print("{s}", .{label});
    for (xs) |x| std.debug.print("{d} ", .{x});
    std.debug.print("\n", .{});
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const input = [_]i32{ 1, 2, 3, 4, 5 };
    var head = try fromArray(a, &input);

    var buf: [16]i32 = undefined;
    var n = toArray(head, &buf);
    printInts("Before: ", buf[0..n]);

    head = reverse(head);

    n = toArray(head, &buf);
    printInts("After:  ", buf[0..n]);

    const expected = [_]i32{ 5, 4, 3, 2, 1 };
    std.debug.assert(n == expected.len);
    for (buf[0..n], expected) |got, exp| std.debug.assert(got == exp);
    std.debug.print("OK — reversed list matches expected\n", .{});
}
